core/services/ccip/orm.go

```diff
@@ -79,82 +67,141 @@ func (o *orm) GetGasPricesByDestChain(ctx context.Context, destChainSelector uin
-	for _, price := range tokenPrices {
+	tokensToUpdate, err := o.pickOnlyRelevantTokensForUpdate(ctx, destChainSelector, tokenPrices, interval)
+	if err != nil || len(tokensToUpdate) == 0 {
+		return 0, err
+	}
+
+	insertData := make([]map[string]interface{}, 0, len(tokensToUpdate))
+	for _, price := range tokensToUpdate {
 		insertData = append(insertData, map[string]interface{}{
 			"chain_selector": destChainSelector,
-			"job_id":         jobId,
 			"token_addr":     price.TokenAddr,
 			"token_price":    price.TokenPrice,
 		})
 	}
 
-	// using statement_timestamp() to make testing easier
-	stmt := `INSERT INTO ccip.observed_token_prices (chain_selector, job_id, token_addr, token_price, created_at)
-		VALUES (:chain_selector, :job_id, :token_addr, :token_price, statement_timestamp());`
-	_, err := o.ds.NamedExecContext(ctx, stmt, insertData)
+	stmt := `INSERT INTO ccip.observed_token_prices (chain_selector, token_addr, token_price, updated_at)
+		VALUES (:chain_selector, :token_addr, :token_price, statement_timestamp())
+		ON CONFLICT (token_addr, chain_selector) 
+		DO UPDATE SET token_price = EXCLUDED.token_price, updated_at = EXCLUDED.updated_at;`
+	result, err := o.ds.NamedExecContext(ctx, stmt, insertData)
 	if err != nil {
-		err = fmt.Errorf("error inserting token prices for job %d: %w", jobId, err)
+		return 0, fmt.Errorf("error inserting token prices %w", err)
 	}
-
-	return err
+	return result.RowsAffected()
 }
 
-func (o *orm) ClearGasPricesByDestChain(ctx context.Context, destChainSelector uint64, expireSec int) error {
-	stmt := `DELETE FROM ccip.observed_gas_prices WHERE chain_selector = $1 AND created_at < (statement_timestamp() - $2 * interval '1 second')`
+// pickOnlyRelevantTokensForUpdate returns only tokens that need to be updated. Multiple jobs can be updating the same tokens,
+// in order to reduce table locking and redundant upserts we start with reading the table and checking which tokens are eligible for update.
+// A token is eligible for update when time since last update is greater than the interval.
+func (o *orm) pickOnlyRelevantTokensForUpdate(
+	ctx context.Context,
+	destChainSelector uint64,
+	tokenPrices []TokenPrice,
+	interval time.Duration,
+) ([]TokenPrice, error) {
+	tokenPricesByAddress := toTokensByAddress(tokenPrices)
+
+	stmt := `
+		SELECT 
+		    token_addr
+		FROM ccip.observed_token_prices
+		WHERE 
+		    chain_selector = $1
+			and token_addr = any($2)
+			and updated_at >= statement_timestamp() - $3::interval
+	`
+
+	pgInterval := fmt.Sprintf("%d milliseconds", interval.Milliseconds())
+	args := []interface{}{destChainSelector, tokenAddrsToBytes(tokenPricesByAddress), pgInterval}
+	var dbTokensToIgnore []string
+	if err := o.ds.SelectContext(ctx, &dbTokensToIgnore, stmt, args...); err != nil {
+		return nil, err
+	}
 
-	_, err := o.ds.ExecContext(ctx, stmt, destChainSelector, expireSec)
-	return err
+	tokensToIgnore := mapset.NewThreadUnsafeSet[string](dbTokensToIgnore...)
```

Comment:
Is the `UnsafeSet` intentional?

Reply:
+1, is there a reason to use `NewThreadUnsafeSet` vs alternatives?

Reply:
Yeah, mapset default implementation is threadsafe which adds locking. It's not needed here, because we use that set only within this function. Locking comes with overhead so I've decided to use the implementation that doesn't have any locks under hood. However, if it raises questions I can either use locking implementation or just use a regular Go map

Reply:
```go
// NewThreadUnsafeSet creates and returns a new set with the given elements.
// Operations on the resulting set are not thread-safe.
```